### PR TITLE
fix: detect type for URLs with query parameter or fragment identifier

### PR DESCRIPTION
**Pull request by devoto13**
Closes #3497

### CODE REVIEW

File: lib/middleware/karma.js [hunk 1]
```diff
@@ -164,10 +162,14 @@ function createKarmaMiddleware (
           const scriptTags = []
           for (const file of files.included) {
             let filePath = file.path
-            const fileType = file.type || path.extname(filePath).substring(1)
-
-            if (helper.isDefined(fileType) && !FILE_TYPES.includes(fileType)) {
-              log.warn(`Invalid file type (${fileType}), defaulting to js.`)
+            const fileType = file.type || file.detectType()
+
+            if (!FILE_TYPES.includes(fileType)) {
```
Comment (devoto13) [inline lib/middleware/karma.js:167]: `helper.isDefined(fileType)` always returns `true` with current code, that's why it was removed.

File: lib/file.js [hunk 2]
```diff
@@ -29,6 +31,14 @@ class File {
     this.isBinary = isBinary === undefined ? null : isBinary
   }
 
+  /**
+   * Detect type from the file extension.
+   * @returns {string} detected file type of empty string
```
Comment (johnjbarton) [inline lib/file.js:36]: maybe "or empty string"

File: lib/url.js [hunk 3]
```diff
@@ -1,15 +1,27 @@
...
   constructor (path, type) {
     this.path = path
+    this.originalPath = path
     this.type = type
     this.isUrl = true
   }
 
+  /**
+   * Detect type from the file extension in the path part of the URL.
+   * @returns {string} detected file type of empty string
```
Comment (johnjbarton) [inline lib/url.js:19]: of -> or

File: lib/middleware/karma.js [hunk 4]
```diff
@@ -164,10 +162,14 @@ function createKarmaMiddleware (
...
+              log.warn(
+                `Invalid file type (${fileType}), defaulting to js.\n` +
```
Comment (johnjbarton) [inline lib/middleware/karma.js:169]: Maybe ($fileType || 'filename has no .extension')

File: lib/file.js (hunk 2 above)
Comment (devoto13) [reply lib/file.js]: Oops, fixed

File: lib/middleware/karma.js (hunk 4 above)
Comment (devoto13) [reply lib/middleware/karma.js]: Not sure if it really makes the warning more understandable... Before it was simply saying that `$fileType` is an invalid file type, but now it also introduces a relation between file type and file extension without really explaining it.

How about having different messages depending on whether it was an invalid value in the configuration (this should probably be changed to an error, but in a separate PR) vs Karma failed to guess a valid type from the file extension? See updated PR.

My logic with the error message is to say:
- what happened? - Karma failed to guess a file type
- how to fix it? - specify file type in the configuration file
- don't understand what is it about? - read the documentation link that explains it

Comment (devoto13) [reply lib/middleware/karma.js]: And my assumption here is that normally people want to specify the `type` in the configuration, not change the extension of the file. E.g. testing `.vue` single file components or https://github.com/karma-runner/karma/pull/3491#issuecomment-621737725.

File: lib/middleware/karma.js [hunk 5]
```diff
@@ -164,10 +162,18 @@ function createKarmaMiddleware (
...
-              log.warn(`Invalid file type (${fileType}), defaulting to js.`)
+            if (!FILE_TYPES.includes(fileType)) {
+              if (file.type == null) {
+                log.warn(
+                  `Unable to determine file type from the file extension, defaulting to js.\n` +
+                  `  To silence the warning specify a valid type for ${file.originalPath} in the configuration file.\n` +
+                  `  See http://karma-runner.github.io/latest/config/files.html`
+                )
+              } else {
+                log.warn(`Invalid file type (${file.type}), defaulting to js.`)
```
Comment (johnjbarton) [inline lib/middleware/karma.js:175]: I just think the message may be confusing in the case `file.type === ''`:
  Invalid file type (), defaulting to js

Comment (devoto13) [reply lib/middleware/karma.js]: It should be `Invalid file type (empty string), defaulting to js.` now.

**Pull request merged**